File: androlua/src/main/java/cn/vove7/androlua/luabridge/LuaTimerTask.java
package cn.vove7.androlua.luabridge;

import cn.vove7.androlua.luautils.TimerTaskX;
import com.luajava.LuaException;
import com.luajava.LuaObject;
import com.luajava.LuaState;

import cn.vove7.androlua.LuaApp;
import cn.vove7.androlua.luautils.LuaManagerI;

/**
 * LuaTask
 */
public class LuaTimerTask extends TimerTaskX implements LuaRunnableI {
    private LuaState L;
    private LuaManagerI luaManager;

    private String mSrc;
    private LuaThreadHelper threadHelper;

    private Object[] mArg = new Object[0];

    private boolean mEnabled = true;

    private byte[] mBuffer;

    public LuaTimerTask(LuaManagerI luaManager, String src) {
        this(luaManager, src, null);
    }

    public LuaTimerTask(LuaManagerI luaManager, String src, Object[] arg) {
        this.luaManager = luaManager;
        mSrc = src;
        if (arg != null)
            mArg = arg;
        L = LuaApp.getInstance().getLuaHelper().L;
        threadHelper = new LuaThreadHelper(luaManager, L);
    }

    @Override
    public void quit() {
        L.gc(LuaState.LUA_GCCOLLECT, 1);
        System.gc();
        if (isCancelled()) {
            luaManager.handleMessage(LuaManagerI.W, getName() + " already canceled");
            return;
        }
        luaManager.log(getName() + " quit");
        cancel();
    }

    public LuaTimerTask(LuaManagerI luaManager, LuaObject func) throws LuaException {
        this(luaManager, func, null);
    }

    public LuaTimerTask(LuaManagerI luaManager, LuaObject func, Object[] arg) throws LuaException {
        this.luaManager = luaManager;
        if (arg != null)
            mArg = arg;
        mBuffer = func.dump();
        L = LuaApp.getInstance().getLuaHelper().L;
        threadHelper = new LuaThreadHelper(luaManager, L);
    }

    String getName() {
        return "TimerTask " + hashCode() + " ";
    }

    @Override
    public void run() {
        if (!mEnabled) {
            luaManager.log(getName() + "Enabled is false");
            return;
        }
        L.getGlobal("run");
        try {
            if (!L.isNil(-1))
                threadHelper.runFunc("run");
            else {
                if (mBuffer != null)
                    threadHelper.newLuaThread(mBuffer, mArg);
                else
                    threadHelper.newLuaThread(mSrc, mArg);
            }
        } catch (LuaException e) {
            luaManager.handleMessage(LuaManagerI.E,e.getMessage());
            e.printStackTrace();
            quit();
        }
    }

    public void setArg(Object[] arg) {
        mArg = arg;
    }

    public void setArg(LuaObject arg) throws ArrayIndexOutOfBoundsException, LuaException, IllegalArgumentException {
        mArg = arg.asArray();
    }

    public boolean isEnabled() {
        return mEnabled;
    }

    public void setEnabled(boolean enabled) {
        mEnabled = enabled;
    }

    public void set(String key, Object value) throws LuaException {
        L.pushObjectValue(value);
        L.setGlobal(key);
    }

    public Object get(String key) throws LuaException {
        L.getGlobal(key);
        return L.toJavaObject(-1);
    }
}

File: androlua/src/main/java/cn/vove7/androlua/luabridge/LuaTimerTask.kt
package cn.vove7.androlua.luabridge

import cn.vove7.androlua.LuaHelper
import cn.vove7.androlua.luautils.LuaManagerI
import cn.vove7.androlua.luautils.LuaRunnableI
import cn.vove7.androlua.luautils.TimerTaskX
import cn.vove7.common.app.GlobalApp
import cn.vove7.common.executor.OnPrint
import cn.vove7.vtp.log.Vog
import com.luajava.LuaException
import com.luajava.LuaObject
import com.luajava.LuaState

/**
 * LuaTask
 */
class LuaTimerTask : TimerTaskX, LuaRunnableI {
    private var L: LuaState
    private var luaManager: LuaManagerI

    private var mSrc: String? = null
    private var funHelper: LuaFunHelper
    private var luaHelper: LuaHelper

    private var mArg = arrayOf<Any>(0)

    private var mEnabled = true

    private var mBuffer: ByteArray? = null

    val name: String
        get() = "TimerTask " + hashCode() + " "

    @JvmOverloads constructor(luaManager: LuaManagerI, src: String, arg: Array<Any>? = null)
            : this(luaManager, arg) {
        mSrc = src
        if (arg != null)
            mArg = arg
    }

    @Throws(LuaException::class)
    @JvmOverloads constructor(luaManager: LuaManagerI, func: LuaObject, arg: Array<Any>? = null)
            : this(luaManager, arg) {
        mBuffer = func.dump()
    }

    @Throws(LuaException::class)
    constructor(luaManager: LuaManagerI, arg: Array<Any>?) {
        if (arg != null)
            mArg = arg
        this.luaManager = luaManager
        luaHelper = LuaHelper(GlobalApp.APP)
        L = luaHelper.L
        funHelper = LuaFunHelper(luaHelper, L)

        funHelper.copyRuntimeFrom(luaManager.luaState)
    }

    override fun quit() {
        quit(false)
    }

    override fun quit(self: Boolean) {
        Vog.d("quit $this $self")
        L.gc(LuaState.LUA_GCCOLLECT, 1)
//        L.close()
        System.gc()
        if (isCancelled) {
            luaManager.handleMessage(OnPrint.WARN, "$name already canceled")
            return
        }
        cancel()
    }


    override fun run() {
        if (!mEnabled) {
            luaManager.log(name + "Enabled is false")
            return
        }
        L.getGlobal("run")
        try {
            if (!L.isNil(-1))
                funHelper.runFunc("run")
            else {
                if (mBuffer != null)
                    funHelper.newLuaThread(mBuffer, *mArg)
                else
                    funHelper.newLuaThread(mSrc, *mArg)
            }
        } catch (e: LuaException) {
            luaManager.handleMessage(OnPrint.ERROR, e.message ?: "")
            e.printStackTrace()
            quit()
        }

    }

    fun setArg(arg: Array<Any>) {
        mArg = arg
    }

    @Throws(ArrayIndexOutOfBoundsException::class, LuaException::class, IllegalArgumentException::class)
    fun setArg(arg: LuaObject) {
        mArg = arg.asArray()
    }

    override fun isEnabled(): Boolean {
        return mEnabled
    }

    override fun setEnabled(enabled: Boolean) {
        mEnabled = enabled
    }

    @Throws(LuaException::class)
    operator fun set(key: String, value: Any) {
        L.pushObjectValue(value)
        L.setGlobal(key)
    }

    @Throws(LuaException::class)
    operator fun get(key: String): Any? {
        L.getGlobal(key)
        return L.toJavaObject(-1)
    }
}
